sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -777,21 +797,23 @@ public PDone expand(PCollection<MutationGroup> input) {
               .apply("Sample keys", sampler)
               .apply("Keys sample as view", View.asMap());
 
+      TupleTag<Void> mainTag = new TupleTag<>("mainOut");
+      TupleTag<MutationGroup> failedTag = new TupleTag<>("failedMutations");
       // Assign partition based on the closest element in the sample and group mutations.
       AssignPartitionFn assignPartitionFn = new AssignPartitionFn(keySample);
-      serialized
+      PCollectionTuple result = serialized
           .apply("Partition input", ParDo.of(assignPartitionFn).withSideInputs(keySample))
           .setCoder(KvCoder.of(StringUtf8Coder.of(), SerializedMutationCoder.of()))
-          .apply("Group by partition", GroupByKey.create())
-          .apply(
-              "Batch mutations together",
+          .apply("Group by partition", GroupByKey.create()).apply("Batch mutations together",
               ParDo.of(new BatchFn(spec.getBatchSizeBytes(), spec.getSpannerConfig(), schemaView))
-                  .withSideInputs(schemaView))
-          .apply(
-              "Write mutations to Spanner",
-              ParDo.of(new WriteToSpannerFn(spec.getSpannerConfig())));
-      return PDone.in(input.getPipeline());
-
+                  .withSideInputs(schemaView)).apply("Write mutations to Spanner",
+              ParDo.of(new WriteToSpannerFn(spec.getSpannerConfig(), spec.getFailureMode(),
+                  failedTag))
+                  .withOutputTags(mainTag, TupleTagList.of(failedTag)));
+      PCollection<MutationGroup> failedMutations = result.get(failedTag);
+      failedMutations.setCoder(SerializableCoder.of(MutationGroup.class));
```

Comment:
It's unfortunate that we can't directly use MutationGroupEncoder as a Coder, since it requires a Schema known at construction time...

Reply:
Agree, if you have ideas how to improve this I'll be happy to help